pulse backend string model parser

DanPuzzuoli opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Closes #53.

Bring string model parser from Aer into dynamics. Currently a work in progress as I attempt to reduce the original number of files and lines of code involved.

### Details and comments

DanPuzzuoli commented:
This PR is ready for review. I think it adds the "smallest usable unit" that makes sense for parsing the pulse backend model strings.

All non-test files added are in `qiskit_dynamics/pulse/string_model_parser`. The main function, and probably the only one that should be used, is `string_model_parser.parse_hamiltonian_dict`, which takes in the backend hamiltonian dictionary, and returns:
- The static hamiltonian,
- A hamiltonian operators (i.e. the terms with time-dependent coefficients in the form of channels/signals), and
- A corresponding list of channel labels for the terms with time-dependent coefficients.

In subsequent developments, the operators can be used to construct a model or solver in dynamics, and the list of channels will be used in the schedule -> signal conversion routine put the output signals in the right order.

Some notes:
- The `legacy_parser.py` file compartmentalizes the actual string parsing part from the Aer pulse simulator, which is untested and that is treated as a black box. At this stage I think there are diminishing returns to trying to understand/test it, and putting it in this file is an attempt to make it clear that we are purposefully compartmentalizing it.
- I've tried to document the expected input format in the doc string for `parse_hamiltonian_dict`. I think this doc string is important to review.
- I've put everything into it's own subfolder since it consists of several files but there's really only one meaningful function coming out of it. I'd be interested in comment on this structure as well.

brosand reviewed (commented):
LGTM, happy to have a longer discussion to be more detailed, but nothing jumps out.

brosand commented on qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:
```diff
@@ -0,0 +1,305 @@
+                if channel_str[1] == "{":
+                    if not channel_str[-1] == "}":
+                        raise QiskitError(malformed_text)
+                # otherwise verify that the remainder of terms are only contains digits
```
verify that the remainder of the terms only contain digits

DanPuzzuoli replied:
In this case they're not digits, they'll be a string of characters and refer to the summation index. I think adding a check for this may be more complicated than its worth.

brosand replied:
yes sorry by digits I meant non-weird chars, but agreed will resolve

brosand commented on qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:
```diff
@@ -0,0 +1,305 @@
+            # Verify either that: all remaining characters are digits, or,
+            # if term starts with _SUM[ and ends with ], all remaining characters
+            # are either digits, or starts and ends with {}
+            if term[-1] == "]" and len(term) > 5 and term[:5] == "_SUM[":
```
Shouldn't we do checks like these even on strings without channels? (specifically line 280)

DanPuzzuoli replied:
It appears all of these checks are based around validating channel specification - nothing else is being checked here. I think I just did this here because the legacy parser has some of its own checking for the validity of the operator part of the string, but has less checking for the validity of the channel part.

brosand commented on test/dynamics/pulse/test_string_model_parser.py:
```diff
@@ -0,0 +1,422 @@
+class TestHamiltonianPreParseExceptions(QiskitDynamicsTestCase):
+    """Tests for preparse formatting exceptions."""
+
+    def test_no_h_str(self):
```
Super unimportant but just to cover edge case of `h_str= ""`, add a line for that as well, since we have the code for it above. (Let me know if this is too unimportant, still trying to figure out best practices here)

DanPuzzuoli replied:
Added two additional checks to this test to verify that the same error gets raised if the argument is: `{'h_str': []}` or `{'h_str': ['']}`.

This is good to point - obviously we can't be overly exhaustive of all possible validation issues but it's good to make sure nothing breaks here.

brosand commented on test/dynamics/pulse/test_string_model_parser.py:
```diff
@@ -0,0 +1,422 @@
+            hamiltonian_pre_parse_exceptions({"h_str": ["a * X0|||D0"]})
+        self.assertTrue("requires non-empty 'qub'" in str(qe.exception))
+
+    def test_too_many_vertical_bars(self):
```
Add a test for too few vertical bars (1)

DanPuzzuoli replied:
Added below this one.

brosand commented on qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:
```diff
@@ -0,0 +1,305 @@
+    Raises:
+        QiskitError: If some part of the Hamiltonian dictionary is unsupported or invalid.
+    """
+
```
Some other ideas for preparsing if we want to (maybe unnecessary)

- check if variables exist which aren't defined (probably handled by exceptions from the evaluation).
- Check for non-alphanumeric characters that don't have predefined uses (unless those are fine as variable names)

DanPuzzuoli replied:
These are good ideas, though interestingly these problems are actually automatically handled gracefully by virtue of how these expressions are evaluated. Since `exec` is called on the coefficient strings, if a variable isn't defined a `NameError` is raised for that variable name. I think the same applies to the second point, and the legacy parser will handle undefined operator names (via calls to `operator_from_string`).

brosand replied:
great this was my hope

brosand reviewed (commented):
Looking good, mostly just left a couple small suggestions for testing and parsing, and a couple small typos.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:
```diff
@@ -0,0 +1,174 @@
+    out = single_operator_from_string(op_label, subsystem_dims[subsystem_index])
+
+    # sort subsystem labels and dimensions
+    sorted_subsystem_keys, sorted_subsystem_dims = zip(
```
Basically this line is set up to sort two lists simultaneously, according to the ordering of the first list. Working inwards out:
1. The first zip combines the list into a single iterator whose elements are tuples with one element from each list (in the original order).
2. `sorted` then sorts this new iterator according to the first element in the tuples (the first list).
3. I'm not fully sure of exactly what this syntax implies, but `zip(*x)` for `x = [(a1, b1), (a2, b2), ... , (ak, bk)]` returns `[a1, ..., ak], [b1, ..., bk]`. (i.e. now that things are sorted, we split the zipped thing back into individual lists.)

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:
```diff
@@ -0,0 +1,174 @@
+        QiskitError: If op_label doesn't correspond to a known operator.
+    """
+
+    op_func = __operdict.get(op_label, None)
```
Moved it above.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:
```diff
@@ -0,0 +1,174 @@
+    return np.conjugate(np.transpose(op))
+
+
+def apply_func(name: str, op: np.ndarray) -> np.ndarray:
```
This is what existed in the original pulse simulator string parser, so just brought that over. The parser still requires this format so I will stick with this and we can add things over time if necessary.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:
```diff
@@ -0,0 +1,305 @@
+    If ``None``, all subsystems are kept. If ``subsystem_list`` is specified, then terms
+    including subsystems not in the list will be ignored.
+
+    Entires in the list ``hamiltonian_dict['h_str']`` must be formated as a product of
```
Done.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:
```diff
@@ -0,0 +1,305 @@
+    If ``None``, all subsystems are kept. If ``subsystem_list`` is specified, then terms
+    including subsystems not in the list will be ignored.
+
+    Entires in the list ``hamiltonian_dict['h_str']`` must be formated as a product of
```
Done.

DanPuzzuoli commented:
Not sure why I can't reply to the one comment about adding the output to one of the examples in the docs. 

I'm kind of on the fence about this - two of the outputs of the function consist of arrays and so showing them isn't necessarily too informative. Given that this function is primarily for internal use anyway I'm leaning towards not adding this.

brosand commented:
> Not sure why I can't reply to the one comment about adding the output to one of the examples in the docs.
> 
> I'm kind of on the fence about this - two of the outputs of the function consist of arrays and so showing them isn't necessarily too informative. Given that this function is primarily for internal use anyway I'm leaning towards not adding this.

Not sure why I can't reply to this one either haha, but this makes sense,

brosand commented on test/dynamics/pulse/test_string_model_parser.py:
```diff
@@ -54,6 +62,13 @@ def test_too_many_vertical_bars(self):
         self.assertTrue("does not conform" in str(qe.exception))
 
+    def test_single_vertical_bars(self):
+        """Test that too many vertical bars raises error."""
```
should be "test that too few vertical bars"

brosand reviewed (dismissed):
LGTM, thanks for doing those small changes

chriseclectic commented on qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:
```diff
@@ -0,0 +1,174 @@
+__funcdict = {"dag": dag}
+
+
+def apply_func(name: str, op: np.ndarray) -> np.ndarray:
```
This function seems sus. You could just replace it in the parser code (which is the only spot it looks to be used) with something like 
```python
elif token.type in ["Func", "Ext"]:
     if token.name == "dag":
         stack.append(np.conjugate(np.transpose(stack.pop(-1)))
     else:
          stack.append(stack.pop(-1))
```
(possibly something even simpler if you know whats going on in that code block)

DanPuzzuoli replied:
yeah you're right - I think maybe it was like this in the original pulse simulator to enable further expansion of functionality later, but for now it's pointless. I've implemented your suggestion and removed `apply_func` and related things, though changed the else statement to raise an error.

chriseclectic commented on qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:
```diff
@@ -0,0 +1,306 @@
+    subsystem_dims = {int(qubit): qub_dict[int(qubit)] for qubit in subsystem_list}
+
+    # Parse the Hamiltonian
+    system = legacy_parser(
```
No need for this function, it could just be `system = HarmiltonianParser(args).parse(other_args)`

DanPuzzuoli replied:
Unless you think otherwise I'd rather keep the `legacy_parser` function even though this is all it does as its the preferred interface (a function call).

chriseclectic commented on qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:
```diff
@@ -0,0 +1,355 @@
+)
+
+
+class HamiltonianParser:
```
Might also want to call this `_HamiltonianParser` just to add further indication this shouldn't be used and is internal only

DanPuzzuoli replied:
No reason - this is part of the "legacy" aspect of it and why I wrote the `legacy_parser` function. Turning it directly into a function was non-trivial given the handling of internal state. The `legacy_parser` function was added to wrap this in the preferred standard function interface.

chriseclectic reviewed (changes requested):
Since majority of this is intended as internal code hidden from users i would recommend adding some comments to that effect in the module docstrings, and renaming most of the classes and functions to start with an underscore to further indicate they are internal methods that should not be used directly

DanPuzzuoli commented on qiskit_dynamics/pulse/__init__.py:
```diff
@@ -75,6 +75,8 @@
 """
 
 from .pulse_to_signals import InstructionToSignals
+from .string_model_parser.string_model_parser import parse_hamiltonian_dict
```
You're right, I'll remove this.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:
```diff
@@ -0,0 +1,355 @@
+from .operator_from_string import operator_from_string, apply_func
+
+
+def legacy_parser(operator_str, subsystem_dims, subsystem_list):
```
I'm using "legacy" here to imply that it's a non-trivial piece of code that is being preserved but not "maintained" per se. I like the usage of "legacy" to imply this - I think the way you're interpreting it is basically synonymous with "expected to be deprecated". 

That all being said I'll change this if it's a sticking point for you.

DanPuzzuoli replied:
Updated the function to have type hints.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:
```diff
@@ -0,0 +1,355 @@
+
+Token = namedtuple("Token", ("type", "name"))
+
+str_elements = OrderedDict(
```
Moved this into the class.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:
```diff
@@ -0,0 +1,355 @@
+        """
+        self.h_str = h_str
+        self.subsystem_dims = {int(label): int(dim) for label, dim in subsystem_dims.items()}
+        self.__td_hams = []
```
No reason, it was in the original code. I've changed it to just `td_hams`.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:
```diff
@@ -0,0 +1,355 @@
+    """Function wrapper for legacy parsing object."""
+
+    system = HamiltonianParser(h_str=operator_str, subsystem_dims=subsystem_dims)
+    system.parse(subsystem_list)
```
Changed it to return as suggested and simplified `legacy_parser` function to one line.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:
```diff
@@ -0,0 +1,355 @@
+        """Return Hamiltonian terms."""
+        return self.__tc_hams + self.__td_hams
+
+    def parse(self, qubit_list=None):
```
You're right - just removed the `compiled` property and made them local variables.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:
```diff
@@ -0,0 +1,355 @@
+        self.__str2qopr = {}
+
+    @property
+    def compiled(self):
```
Yep - removed.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:
```diff
@@ -0,0 +1,174 @@
+}
+
+
+def single_operator_from_string(op_label: str, dim: int) -> np.ndarray:
```
yep - removed.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:
```diff
@@ -0,0 +1,174 @@
+# that they have been altered from the originals.
+# pylint: disable=invalid-name
+
+"""Generate operators from string."""
```
Added a warning to this and every other file in this submodule.

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:
```diff
@@ -0,0 +1,174 @@
+
+    Args:
+        op_label: The string labelling the single system operator.
+        subsystem_index: Index of the subsystem to apply the operator.
```
This subsystem index is just a single `int` indicating the label of the subsystem that the operator being constructed acts on. The `subsystem_dims` argument is a dictionary mapping subsystem labels to dimensions, so that this function knows where and what order to tensor on identities. The subsystem labels in `subsystem_dims` are assumed to be ints by this function.

To address this:
- Made the type hint for `subsystem_labels` more detailed to show that both the keys and labels are ints.
- Changing `subsystem_index` to `subsystem_label` to avoid confusion that "index" might be indexing a list of subsystems (as opposed to specifying a specific label).

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:
```diff
@@ -0,0 +1,174 @@
+
+
+# operator names
+__operdict = {
```
So that it isn't constructed every time the function is called.

DanPuzzuoli commented:
@chriseclectic I think I've addressed most comments with the latest changes. The remaining things to discuss are:
- Usage of the word "legacy". Maybe we can change the file to `regex_parser.py` and this function to `regex_parser` - this will drop the problematic terminology and may be a more informative name (indicating this is where the true "parsing" happens).
- Using underscores for functions/classes that are meant to be internal: I'd prefer not to do this as I feel like it is unnecessary and I don't really like the way it looks. I think the fact that these things aren't being imported into the standard packaging structure/API, along with the warnings in the files, should be sufficient to imply that they are internal?

DanPuzzuoli commented on qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:
```diff
@@ -0,0 +1,174 @@
+        out = np.kron(out, ident(total_dim))
+
+    # tensor identity on left if subsystem_index not the last subsystem
+    if subsystem_location + 1 != len(sorted_subsystem_keys):
```
Changed it to utilize this.

DanPuzzuoli commented:
Closing this PR as it is extremely out of date - will reopen a new one.
